File: Backend/internal/service/threshold_service.go
package service

import (
	"context"
	"fmt"

	"github.com/buonnguwaaa/Remote-Patient-Monitoring/Backend/internal/domain"
	userDomain "github.com/buonnguwaaa/Remote-Patient-Monitoring/Backend/internal/domain/user"
	"github.com/buonnguwaaa/Remote-Patient-Monitoring/Backend/internal/dto"
	"github.com/buonnguwaaa/Remote-Patient-Monitoring/Backend/internal/repository"
	userRepository "github.com/buonnguwaaa/Remote-Patient-Monitoring/Backend/internal/repository/user"
	"github.com/buonnguwaaa/Remote-Patient-Monitoring/Backend/internal/usecase"
	"github.com/buonnguwaaa/Remote-Patient-Monitoring/Backend/internal/util"
)

type thresholdService struct {
	patientRepo   userRepository.PatientRepository
	doctorRepo    userRepository.StaffRepository[userDomain.Doctor]
	thresholdRepo repository.ThresholdRepository
}

type ThresholdService interface {
	CreateThreshold(context.Context, *usecase.CreateThresholdInput) (*dto.ThresholdResponse, error)
	UpdateThreshold(context.Context, *usecase.UpdateThresholdInput) (*dto.ThresholdResponse, error)
	GetThresholds(context.Context, *usecase.GetThresholdsInput) ([]dto.ThresholdResponse, error)
}

func NewThresholdService(patientRepo userRepository.PatientRepository, doctorRepo userRepository.StaffRepository[userDomain.Doctor], thresholdRepo repository.ThresholdRepository) ThresholdService {
	return &thresholdService{
		patientRepo:   patientRepo,
		doctorRepo:    doctorRepo,
		thresholdRepo: thresholdRepo,
	}
}

func (s *thresholdService) CreateThreshold(ctx context.Context, input *usecase.CreateThresholdInput) (*dto.ThresholdResponse, error) {
	patientID, err := util.MustHexToObjectID(input.PatientID)
	if err != nil {
		return nil, err
	}
	existedPatient, err := s.patientRepo.ExistsByIDAndRole(ctx, patientID, userDomain.RolePatient)
	if err != nil || !existedPatient {
		return nil, fmt.Errorf("user not found or not patient")
	}

	doctorID, err := util.MustHexToObjectID(input.DoctorID)
	if err != nil {
		return nil, err
	}
	existedDoctor, err := s.doctorRepo.ExistsByIDAndRole(ctx, doctorID, userDomain.RoleDoctor)
	if err != nil || !existedDoctor {
		return nil, fmt.Errorf("user not found or not doctor")
	}

	threshold := &domain.Threshold{
		PatientID:          patientID,
		DoctorID:           doctorID,
		TemperatureMin:     input.TemperatureMin,
		TemperatureMax:     input.TemperatureMax,
		HeartRateMin:       input.HeartRateMin,
		HeartRateMax:       input.HeartRateMax,
		RespiratoryRateMin: input.RespiratoryRateMin,
		RespiratoryRateMax: input.RespiratoryRateMax,
		SpO2Min:            input.SpO2Min,
		SysMin:             input.SysMin,
		SysMax:             input.SysMax,
		DiaMin:             input.DiaMin,
		DiaMax:             input.DiaMax,
		GlucoseMin:         input.GlucoseMin,
		GlucoseMax:         input.GlucoseMax,
		EffectiveFrom:      input.EffectiveFrom,
		EffectiveTo:        input.EffectiveTo,
	}

	inserted, err := s.thresholdRepo.Create(ctx, threshold)
	if err != nil {
		return nil, err
	}

	return &dto.ThresholdResponse{
		ID:                 inserted.ID.Hex(),
		PatientID:          inserted.PatientID.Hex(),
		DoctorID:           inserted.DoctorID.Hex(),
		TemperatureMin:     inserted.TemperatureMin,
		TemperatureMax:     inserted.TemperatureMax,
		HeartRateMin:       inserted.HeartRateMin,
		HeartRateMax:       inserted.HeartRateMax,
		RespiratoryRateMin: inserted.RespiratoryRateMin,
		RespiratoryRateMax: inserted.RespiratoryRateMax,
		SpO2Min:            inserted.SpO2Min,
		SysMin:             inserted.SysMin,
		SysMax:             inserted.SysMax,
		DiaMin:             inserted.DiaMin,
		DiaMax:             inserted.DiaMax,
		GlucoseMin:         inserted.GlucoseMin,
		GlucoseMax:         inserted.GlucoseMax,
		EffectiveFrom:      inserted.EffectiveFrom,
		EffectiveTo:        inserted.EffectiveTo,
		CreatedAt:          inserted.CreatedAt,
		UpdatedAt:          inserted.UpdatedAt,
	}, nil
}

func (s *thresholdService) GetThresholds(ctx context.Context, input *usecase.GetThresholdsInput) ([]dto.ThresholdResponse, error) {
	filter := repository.ThresholdFilter{
		PatientID: input.PatientID,
		DoctorID:  input.DoctorID,
		IsLatest:  input.IsLatest,
	}

	thresholds, err := s.thresholdRepo.FindWithFilter(ctx, filter)
	if err != nil {
		return nil, err
	}

	var resp []dto.ThresholdResponse
	for _, t := range thresholds {
		resp = append(resp, dto.ThresholdResponse{
			ID:                 t.ID.Hex(),
			PatientID:          t.PatientID.Hex(),
			DoctorID:           t.DoctorID.Hex(),
			TemperatureMin:     t.TemperatureMin,
			TemperatureMax:     t.TemperatureMax,
			HeartRateMin:       t.HeartRateMin,
			HeartRateMax:       t.HeartRateMax,
			RespiratoryRateMin: t.RespiratoryRateMin,
			RespiratoryRateMax: t.RespiratoryRateMax,
			SpO2Min:            t.SpO2Min,
			SysMin:             t.SysMin,
			SysMax:             t.SysMax,
			DiaMin:             t.DiaMin,
			DiaMax:             t.DiaMax,
			GlucoseMin:         t.GlucoseMin,
			GlucoseMax:         t.GlucoseMax,
			EffectiveFrom:      t.EffectiveFrom,
			EffectiveTo:        t.EffectiveTo,
			CreatedAt:          t.CreatedAt,
			UpdatedAt:          t.UpdatedAt,
		})
	}

	return resp, nil
}

func (s *thresholdService) UpdateThreshold(ctx context.Context, input *usecase.UpdateThresholdInput) (*dto.ThresholdResponse, error) {
	id, err := util.MustHexToObjectID(input.ID)
	if err != nil {
		return nil, err
	}

	newThreshold := &domain.Threshold{
		ID:                 id,
		TemperatureMin:     input.TemperatureMin,
		TemperatureMax:     input.TemperatureMax,
		HeartRateMin:       input.HeartRateMin,
		HeartRateMax:       input.HeartRateMax,
		RespiratoryRateMin: input.RespiratoryRateMin,
		RespiratoryRateMax: input.RespiratoryRateMax,
		SpO2Min:            input.SpO2Min,
		SysMin:             input.SysMin,
		SysMax:             input.SysMax,
		DiaMin:             input.DiaMin,
		DiaMax:             input.DiaMax,
		GlucoseMin:         input.GlucoseMin,
		GlucoseMax:         input.GlucoseMax,
		EffectiveFrom:      input.EffectiveFrom,
		EffectiveTo:        input.EffectiveTo,
	}

	updated, err := s.thresholdRepo.Update(ctx, newThreshold)
	if err != nil {
		return nil, err
	}

	return &dto.ThresholdResponse{
		ID:                 updated.ID.Hex(),
		PatientID:          updated.PatientID.Hex(),
		DoctorID:           updated.DoctorID.Hex(),
		TemperatureMin:     updated.TemperatureMin,
		TemperatureMax:     updated.TemperatureMax,
		HeartRateMin:       updated.HeartRateMin,
		HeartRateMax:       updated.HeartRateMax,
		RespiratoryRateMin: updated.RespiratoryRateMin,
		RespiratoryRateMax: updated.RespiratoryRateMax,
		SpO2Min:            updated.SpO2Min,
		SysMin:             updated.SysMin,
		SysMax:             updated.SysMax,
		DiaMin:             updated.DiaMin,
		DiaMax:             updated.DiaMax,
		GlucoseMin:         updated.GlucoseMin,
		GlucoseMax:         updated.GlucoseMax,
		EffectiveFrom:      updated.EffectiveFrom,
		EffectiveTo:        updated.EffectiveTo,
		CreatedAt:          updated.CreatedAt,
		UpdatedAt:          updated.UpdatedAt,
	}, nil
}

File: Frontend/web/src/services/thresholdService.ts
import api from "./api";

export interface ThresholdPayload {
  patientId: string;
  doctorId: string;
  temperatureMin: number;
  temperatureMax: number;
  heartRateMin: number;
  heartRateMax: number;
  respiratoryRateMin: number;
  respiratoryRateMax: number;
  spo2Min: number;
  sysMin: number;
  sysMax: number;
  diaMin: number;
  diaMax: number;
  glucoseMin: number | null;
  glucoseMax: number | null;
  effectiveFrom: string;
  effectiveTo: string | null;
}

export interface ThresholdRecord {
  id: string;
  patientId: string;
  doctorId: string;
  temperatureMin: number;
  temperatureMax: number;
  heartRateMin: number;
  heartRateMax: number;
  respiratoryRateMin: number;
  respiratoryRateMax: number;
  spo2Min: number;
  sysMin: number;
  sysMax: number;
  diaMin: number;
  diaMax: number;
  glucoseMin: number | null;
  glucoseMax: number | null;
  effectiveFrom: string;
  effectiveTo: string | null;
  createdAt: string;
  updatedAt: string;
}

interface ThresholdApiResponse {
  id: string;
  patientId: string;
  doctorId: string;
  temperatureMin?: number;
  temperatureMax?: number;
  heartRateMin?: number;
  heartRateMax?: number;
  respiratoryRateMin?: number;
  respiratoryRateMax?: number;
  spo2Min?: number;
  sysMin?: number;
  sysMax?: number;
  diaMin?: number;
  diaMax?: number;
  glucoseMin?: number | null;
  glucoseMax?: number | null;
  effectiveFrom: string;
  effectiveTo?: string | null;
  createdAt: string;
  updatedAt: string;
}

const mapThreshold = (item: ThresholdApiResponse): ThresholdRecord => ({
  id: item.id,
  patientId: item.patientId,
  doctorId: item.doctorId,
  temperatureMin: item.temperatureMin ?? 0,
  temperatureMax: item.temperatureMax ?? 0,
  heartRateMin: item.heartRateMin ?? 0,
  heartRateMax: item.heartRateMax ?? 0,
  respiratoryRateMin: item.respiratoryRateMin ?? 0,
  respiratoryRateMax: item.respiratoryRateMax ?? 0,
  spo2Min: item.spo2Min ?? 0,
  sysMin: item.sysMin ?? 0,
  sysMax: item.sysMax ?? 0,
  diaMin: item.diaMin ?? 0,
  diaMax: item.diaMax ?? 0,
  glucoseMin: item.glucoseMin ?? null,
  glucoseMax: item.glucoseMax ?? null,
  effectiveFrom: item.effectiveFrom,
  effectiveTo: item.effectiveTo ?? null,
  createdAt: item.createdAt,
  updatedAt: item.updatedAt,
});

const buildPayload = (payload: ThresholdPayload) => ({
  patientId: payload.patientId,
  doctorId: payload.doctorId,
  temperatureMin: payload.temperatureMin,
  temperatureMax: payload.temperatureMax,
  heartRateMin: payload.heartRateMin,
  heartRateMax: payload.heartRateMax,
  respiratoryRateMin: payload.respiratoryRateMin,
  respiratoryRateMax: payload.respiratoryRateMax,
  spo2Min: payload.spo2Min,
  sysMin: payload.sysMin,
  sysMax: payload.sysMax,
  diaMin: payload.diaMin,
  diaMax: payload.diaMax,
  glucoseMin: payload.glucoseMin,
  glucoseMax: payload.glucoseMax,
  effectiveFrom: payload.effectiveFrom,
  effectiveTo: payload.effectiveTo,
});

export const getThresholds = async (params: {
  patientId?: string;
  doctorId?: string;
  latest?: boolean;
}) => {
  const response = await api.get<{ data: ThresholdApiResponse[] | null }>("/thresholds", {
    params: {
      patientId: params.patientId,
      doctorId: params.doctorId,
      latest: params.latest ? "true" : undefined,
    },
  });

  return (response.data.data || []).map(mapThreshold);
};

export const createThreshold = async (payload: ThresholdPayload) => {
  const response = await api.post<{ data: ThresholdApiResponse }>("/thresholds", buildPayload(payload));
  return mapThreshold(response.data.data);
};

export const updateThreshold = async (id: string, payload: ThresholdPayload) => {
  const response = await api.patch<{ data: ThresholdApiResponse }>(`/thresholds/${id}`, {
    temperatureMin: payload.temperatureMin,
    temperatureMax: payload.temperatureMax,
    heartRateMin: payload.heartRateMin,
    heartRateMax: payload.heartRateMax,
    respiratoryRateMin: payload.respiratoryRateMin,
    respiratoryRateMax: payload.respiratoryRateMax,
    spo2Min: payload.spo2Min,
    sysMin: payload.sysMin,
    sysMax: payload.sysMax,
    diaMin: payload.diaMin,
    diaMax: payload.diaMax,
    glucoseMin: payload.glucoseMin,
    glucoseMax: payload.glucoseMax,
    effectiveFrom: payload.effectiveFrom,
    effectiveTo: payload.effectiveTo,
  });

  return mapThreshold(response.data.data);
};

export const deleteThreshold = async (id: string) => {
  await api.delete(`/thresholds/${id}`);
};
